dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceCodeRepoInterface.java

```diff
@@ -256,83 +264,102 @@ public Workflow getWorkflow(String repositoryId, Optional<Workflow> existingWork
+        // If no tags or workflow versions, have no metadata
+        if (entry.getWorkflowVersions().isEmpty()) {
             return entry;
         }
 
-        // Determine the file path of the descriptor
-        String filePath = null;
-        Set<SourceFile> sourceFiles = null;
-
-        // If entry is a tool
         if (entry instanceof Tool) {
-            // If no tags exist on quay
-            if (entry.getWorkflowVersions().isEmpty()) {
-                return entry;
-            }
-
-            // Find filepath to parse
-            for (Tag tag : ((Tool)entry).getWorkflowVersions()) {
-                if (tag.getReference() != null && tag.getReference().equals(branch)) {
-                    sourceFiles = tag.getSourceFiles();
-                    if (type == DescriptorLanguage.CWL) {
-                        filePath = tag.getCwlPath();
-                        version = tag;
-                    } else if (type == DescriptorLanguage.WDL) {
-                        filePath = tag.getWdlPath();
-                        version = tag;
-                    } else {
-                        throw new UnsupportedOperationException("tool is not a CWL or WDL file");
-                    }
+            Tool tool = (Tool)entry;
+            tool.getWorkflowVersions().forEach(tag -> {
+                String filePath;
+                if (type == DescriptorLanguage.CWL) {
+                    filePath = tag.getCwlPath();
+                } else if (type == DescriptorLanguage.WDL) {
+                    filePath = tag.getWdlPath();
+                } else {
+                    throw new UnsupportedOperationException("tool is not a CWL or WDL file");
                 }
-            }
+                updateVersionMetadata(filePath, tag, type, repositoryId);
+            });
         }
-
-        // If entry is a workflow
-
         if (entry instanceof Workflow) {
-            // Find filepath to parse
-            for (WorkflowVersion workflowVersion : ((Workflow)entry).getWorkflowVersions()) {
-                if (workflowVersion.getReference().equals(branch)) {
-                    filePath = workflowVersion.getWorkflowPath();
-                    sourceFiles = workflowVersion.getSourceFiles();
-                    version = workflowVersion;
-                }
+            Workflow workflow = (Workflow)entry;
+            workflow.getWorkflowVersions().forEach(workflowVersion -> {
+                String filePath = workflowVersion.getWorkflowPath();
+                updateVersionMetadata(filePath, workflowVersion, type, repositoryId);
+            });
+        }
+        return entry;
+    }
+
+    /**
+     * Sets the default version if there isn't already one present.
+     * This is required because entry-level metadata depends on the default version
+     *
+     * @param entry
+     * @param repositoryId
+     */
+    public void setDefaultBranch(Entry entry, String repositoryId) {
```

Comment:
I think the name of the function should be something like setDefaultBranchIfNotSet or something else where you can figure out the behavior without having to read the JavaDoc.